### PR TITLE
Optimization for caches + clearGraphCache()

### PR DESCRIPTION
**Pull request by mikaelforeseeti**
A minor optimization to the graph caching, fewer temporary methods. Also added the clearGraphCache() method to attacksteps.

### CODE REVIEW

File: malcompiler-lib/src/main/java/org/mal_lang/compiler/lib/securicad/AttackStepGenerator.java [hunk 1]
```diff
@@ -77,21 +77,48 @@ protected void generate(TypeSpec.Builder parentBuilder, Asset asset, AttackStep
...
+      TypeSpec.Builder parentBuilder,
+      boolean hasChildCache,
+      String childCacheName,
+      boolean hasParentCache,
+      String parentCacheName) {
+    MethodSpec.Builder builder = MethodSpec.methodBuilder("clearGraphCache");
+    builder.addAnnotation(Override.class);
+    builder.addModifiers(Modifier.PUBLIC);
+    if (hasChildCache) {
+      builder.addCode(String.format("%s.clear();\n", childCacheName));
```
Comment (maxwalls) [inline malcompiler-lib/src/main/java/org/mal_lang/compiler/lib/securicad/AttackStepGenerator.java]: Change to `builder.addStatement("%N.clear()", childCacheName)`

File: malcompiler-lib/src/main/java/org/mal_lang/compiler/lib/securicad/AttackStepGenerator.java [hunk 2]
```diff
@@ -77,21 +77,48 @@ protected void generate(TypeSpec.Builder parentBuilder, Asset asset, AttackStep
...
+    }
+    if (hasParentCache) {
+      builder.addCode(String.format("%s.clear();\n", parentCacheName));
```
Comment (maxwalls) [inline malcompiler-lib/src/main/java/org/mal_lang/compiler/lib/securicad/AttackStepGenerator.java]: Change to `builder.addStatement("%N.clear()", parentCacheName)`

File: malcompiler-lib/src/main/java/org/mal_lang/compiler/lib/securicad/ExpressionGenerator.java [hunk 3]
```diff
@@ -61,19 +61,21 @@ protected void createGetAttackStepChildren(
 
     builder.beginControlFlow("if ($N == null)", cacheName);
     if (attackStep.inheritsReaches()) {
-      builder.addStatement("$L = new $T<>(super.getAttackStepChildren())", cacheName, hashSet);
+      builder.addStatement(
+          "HashSet<AttackStep> tmpCache = new $T<>(super.getAttackStepChildren())", hashSet);
```
Comment (maxwalls) [inline malcompiler-lib/src/main/java/org/mal_lang/compiler/lib/securicad/ExpressionGenerator.java]: Parameterize left-hand type, which should be `Map<AttackStep>`

File: malcompiler-lib/src/main/java/org/mal_lang/compiler/lib/securicad/ExpressionGenerator.java [hunk 4]
```diff
@@ -61,19 +61,21 @@ protected void createGetAttackStepChildren(
...
     } else {
-      builder.addStatement("$L = new $T<>()", cacheName, hashSet);
+      builder.addStatement("HashSet<AttackStep> tmpCache = new $T<>()", hashSet);
```
Comment (maxwalls) [inline malcompiler-lib/src/main/java/org/mal_lang/compiler/lib/securicad/ExpressionGenerator.java]: same

File: malcompiler-lib/src/main/java/org/mal_lang/compiler/lib/securicad/ExpressionGenerator.java [hunk 5]
```diff
@@ -61,19 +61,21 @@ protected void createGetAttackStepChildren(
...
     }
     for (StepExpr expr : attackStep.getReaches()) {
       AutoFlow af = new AutoFlow();
       AutoFlow end = generate(af, expr, attackStep.getAsset(), "(null)");
-      end.addStatement("$L.add($L)", cacheName, end.prefix);
+      end.addStatement("tmpCache.add($L)", end.prefix);
       af.build(builder);
     }
+    builder.addCode("$N = Set.copyOf(tmpCache);\n", cacheName);
```
Comment (maxwalls) [inline malcompiler-lib/src/main/java/org/mal_lang/compiler/lib/securicad/ExpressionGenerator.java]: Use `%T` for `Set`, and use `builder.addStatement`, without `;\n` in string

File: malcompiler-lib/src/main/java/org/mal_lang/compiler/lib/securicad/ExpressionGenerator.java [hunk 6]
```diff
@@ -61,19 +61,20 @@ protected void createGetAttackStepChildren(
...
+      builder.addStatement("$T tmpCache = new $T<>()", asSet, hashSet);
     }
     for (StepExpr expr : attackStep.getReaches()) {
       AutoFlow af = new AutoFlow();
       AutoFlow end = generate(af, expr, attackStep.getAsset(), "(null)");
-      end.addStatement("$L.add($L)", cacheName, end.prefix);
+      end.addStatement("tmpCache.add($L)", end.prefix);
       af.build(builder);
     }
+    builder.addStatement("$N = Set.copyOf(tmpCache);", cacheName);
```
Comment (maxwalls) [inline malcompiler-lib/src/main/java/org/mal_lang/compiler/lib/securicad/ExpressionGenerator.java]: Remove semicolon

Comment (maxwalls) [reply malcompiler-lib/src/main/java/org/mal_lang/compiler/lib/securicad/ExpressionGenerator.java]: Parameterize `Set`

File: malcompiler-lib/src/main/java/org/mal_lang/compiler/lib/securicad/ExpressionGenerator.java [hunk 7]
```diff
@@ -61,19 +61,20 @@ protected void createGetAttackStepChildren(
...
+      builder.addStatement("$T tmpCache = new $T<>()", asSet, hashSet);
     }
     for (StepExpr expr : attackStep.getReaches()) {
       AutoFlow af = new AutoFlow();
       AutoFlow end = generate(af, expr, attackStep.getAsset(), "(null)");
-      end.addStatement("$L.add($L)", cacheName, end.prefix);
+      end.addStatement("tmpCache.add($L)", end.prefix);
       af.build(builder);
     }
+    builder.addStatement("$N = Set.copyOf(tmpCache)", cacheName);
```
Comment (maxwalls) [inline malcompiler-lib/src/main/java/org/mal_lang/compiler/lib/securicad/ExpressionGenerator.java]: Parameterize `Set`

File: malcompiler-lib/src/main/java/org/mal_lang/compiler/lib/securicad/AttackStepGenerator.java [hunk 8]
```diff
@@ -77,21 +77,48 @@ protected void generate(TypeSpec.Builder parentBuilder, Asset asset, AttackStep
...
+      boolean hasParentCache,
+      String parentCacheName) {
+    MethodSpec.Builder builder = MethodSpec.methodBuilder("clearGraphCache");
+    builder.addAnnotation(Override.class);
+    builder.addModifiers(Modifier.PUBLIC);
+    if (hasChildCache) {
+      builder.addStatement("$N.clear()", childCacheName);
+    }
+    if (hasParentCache) {
+      builder.addStatement("$N.clear()", parentCacheName);
```
Comment (simonwrede-foreseeti) [inline malcompiler-lib/src/main/java/org/mal_lang/compiler/lib/securicad/AttackStepGenerator.java]: These need to be set to null

**Pull request merged**